backbone.js

```diff
@@ -658,29 +679,33 @@
     destroy: function(options) {
       options = options ? _.clone(options) : {};
       var model = this;
-      var success = options.success;
       var wait = options.wait;
+      var success = options.success;
+      var error = options.error;
+      delete options.success;
+      delete options.error;
 
       var destroy = function() {
         model.stopListening();
         model.trigger('destroy', model, model.collection, options);
       };
 
-      options.success = function(resp) {
-        if (wait) destroy();
-        if (success) success.call(options.context, model, resp, options);
-        if (!model.isNew()) model.trigger('sync', model, resp, options);
-      };
-
-      var xhr = false;
-      if (this.isNew()) {
-        _.defer(options.success);
-      } else {
-        wrapError(this, options);
-        xhr = this.sync('delete', this, options);
-      }
+      var isNew = model.isNew();
+      var promise = isNew ? Promise.resolve(model.attributes) :
+        wrapError(this.sync('delete', this, options), this, options, error);
       if (!wait) destroy();
-      return xhr;
+      return promise.then(function(resp) {
+        var promise = Promise.resolve(model);
+        if (wait) destroy();
+        if (success)
+          promise = Promise.resolve(success.call(options.context, model, resp, options));
+        if (!isNew)
+          promise = promise.then(function(delivered) {
+            model.trigger('sync', model, resp, options);
```

Comment:
This is super important to me, the other day I got bit by an error that was being thrown by a jquery plugin inside a [stickit](https://github.com/NYTimes/backbone.stickit) method that is called by a sync trigger. The error was swallowed deep inside a promise (of my own making) and was very hard to trace. If I had this code the error object would have been propagated up to the user and the user would have probably/hopefully been presented with an error instead of half a page of broken content. Of course the outcome is that I had to work around the buggy jquery plugin (there's no escape from that!) but there would have been a better developer experience at the very least.

So, I wrote a test: 403bc44a7b89c70a49d4c6f12675b06dbc0932e6